ee/desktop/runner/runner.go

```diff
@@ -324,32 +326,85 @@ func (r *DesktopUsersProcessesRunner) Ping() {
 	level.Debug(r.logger).Log("msg", "runner processSpawningEnabled:%s", strconv.FormatBool(enabled))
 }
 
-// writeMenuFile writes menu data to a shared file for user processes to access
-func (r *DesktopUsersProcessesRunner) writeMenuFile(data menu.MenuData) error {
+// writeSharedFile writes data to a shared file for user processes to access
+func (r *DesktopUsersProcessesRunner) writeSharedFile(path string, data any) error {
 	menuBytes, err := json.Marshal(data)
 	if err != nil {
-		return fmt.Errorf("failed to marshal menu json: %w", err)
+		return fmt.Errorf("failed to marshal json: %w", err)
 	}
 
-	menuPath := r.menuPath()
-	statusFile, err := os.Create(menuPath)
+	file, err := os.Create(path)
 	if err != nil {
-		return fmt.Errorf("creating desktop menu file: %w", err)
+		return fmt.Errorf("creating file: %w", err)
 	}
 
-	if err := os.Chmod(menuPath, 0644); err != nil {
+	if err := os.Chmod(path, 0644); err != nil {
 		return fmt.Errorf("os.Chmod: %w", err)
 	}
 
-	defer statusFile.Close()
-	_, err = io.Copy(statusFile, bytes.NewReader(menuBytes))
+	defer file.Close()
+	_, err = file.Write(menuBytes)
 	if err != nil {
-		return fmt.Errorf("writing desktop menu file: %w", err)
+		return fmt.Errorf("writing file: %w", err)
+	}
+
+	return nil
+}
+
+// generateMenuFile generates and writes menu data to a shared file
+func (r *DesktopUsersProcessesRunner) generateMenuFile(data io.Reader) error {
+	// First generate fresh template data to use for parsing
+	v := version.Version()
+	td := &menu.TemplateData{
+		LauncherVersion:  v.Version,
+		LauncherRevision: v.Revision,
+		GoVersion:        v.GoVersion,
+		ServerHostname:   r.hostname,
+	}
+
+	menuDataBytes, err := io.ReadAll(data)
+	if err != nil {
+		return fmt.Errorf("failed to read menu data: %w", err)
+	}
+
+	// Convert the raw JSON to a string and feed it to the parser for template expansion
+	parser := menu.NewTemplateParser(td)
+	parsedMenuDataStr, err := parser.Parse(string(menuDataBytes))
+	if err != nil {
+		return fmt.Errorf("failed to parse menu data: %w", err)
+	}
+
+	// Convert the parsed string back to bytes, which can now be decoded per usual
+	parsedMenuDataBytes := []byte(parsedMenuDataStr)
+
+	var menu menu.MenuData
+	if err := json.NewDecoder(bytes.NewReader(parsedMenuDataBytes)).Decode(&menu); err != nil {
```

Comment:
I'd say lets followup, but I think this is all cleaner if `Parse` works off io readers and writers. There's a lot of extra buffering

Reply:
Cool, I'll add this to a follow up PR!